### PR TITLE
Feature/liquidity pool liquidate

### LINKED ISSUE #115: Margin protocol: liquidity_pool_liquidate
implement `liquidity_pool_liquidate`

### PR DESCRIPTION
**Pull request by zjb0807**
Closes #115

### CODE REVIEW

**Comment by zjb0807** [comment]
Did the same thing as ermalkaleci:
- fix conflicts and CI/CD
- add `ensure_can_withdrawal` in LiquidityPoolManager traits.
It may be difficult to merge.

File: modules/margin-liquidity-pools/src/lib.rs [hunk 1]
```diff
@@ -262,7 +261,7 @@ impl<T: Trait> LiquidityPools<T::AccountId> for Module<T> {
 
 	fn ensure_liquidity(pool_id: Self::LiquidityPoolId, amount: Self::Balance) -> DispatchResult {
 		ensure!(Self::balances(&pool_id) >= amount, Error::<T>::CannotWithdrawAmount);
-		T::PoolManager::ensure_pool_safe_after_withdrawal(pool_id, amount)
+		T::PoolManager::ensure_can_withdrawal(pool_id, amount)
```
Comment (shaopengw) [inline modules/margin-liquidity-pools/src/lib.rs]:
```suggestion
		T::PoolManager::ensure_can_withdraw(pool_id, amount)
```

File: modules/traits/src/lib.rs [hunk 2]
```diff
@@ -55,3 +55,7 @@ pub trait MarginProtocolLiquidityPools<AccountId>: LiquidityPools<AccountId> {
 		leveraged_amount: Balance,
 	) -> bool;
 }
+
+pub trait LaminarTreasry<AccountId> {
```
Comment (shaopengw) [inline modules/traits/src/lib.rs]:
```suggestion
pub trait Treasury<AccountId> {
```

File: runtime/src/lib.rs [hunk 3]
```diff
@@ -513,11 +513,20 @@ impl synthetic_protocol::Trait for Runtime {
 
 type TransactionSubmitter = system::offchain::TransactionSubmitter<(), Runtime, UncheckedExtrinsic>;
 
+// TODO: implement LaminarTreasry
+pub struct MockLaminarTreasry;
```
Comment (shaopengw) [inline runtime/src/lib.rs]:
```suggestion
pub struct MockLaminarTreasury;
```

File: modules/margin-protocol/src/tests.rs [hunk 4]
```diff
@@ -894,6 +894,46 @@ fn liquidity_pool_margin_call_and_become_safe_work() {
 		});
 }
 
+#[test]
+fn liquidity_pool_liquidate_work() {
```
Comment (shaopengw) [inline modules/margin-protocol/src/tests.rs]:
```suggestion
fn liquidity_pool_liquidate_works() {
```

File: modules/margin-protocol/src/lib.rs [hunk 5]
```diff
@@ -442,6 +452,33 @@ impl<T: Trait> Module<T> {
 		}
 		Ok(())
 	}
+
+	fn _liquidity_pool_liquidate(pool: LiquidityPoolId) -> DispatchResult {
+		let (enp, ell) = Self::_enp_and_ell(pool, None, None)?;
+		let need_liquidate = enp <= Self::liquidity_pool_enp_threshold().stop_out.into()
```
Comment (shaopengw) [inline modules/margin-protocol/src/lib.rs]:
```suggestion
		let need_liquidating = enp <= Self::liquidity_pool_enp_threshold().stop_out.into()
```

File: modules/margin-protocol/src/lib.rs [hunk 6]
```diff
@@ -788,6 +825,46 @@ impl<T: Trait> Module<T> {
 			Ok(())
 		}
 	}
+
+	/// Force closure position to liquidate liquidity pool based on opened positions.
+	///
+	/// Return `Ok` if closure success, or `Err` if not.
+	fn _pool_liquidate(pool: LiquidityPoolId, position_id: PositionId) -> DispatchResult {
```
Comment (shaopengw) [inline modules/margin-protocol/src/lib.rs]: The name `_pool_liquidate` vs `_liquidity_pool_liquidate` is confusing.

File: modules/margin-protocol/src/lib.rs [hunk 7]
```diff
@@ -408,10 +412,16 @@ impl<T: Trait> Module<T> {
...
 
 		// Close position as much as possible
-		// TODO: print error log
-		<PositionsByTrader<T>>::iter_prefix(who).for_each(|user_position_ids| {
-			user_position_ids.iter().for_each(|position_id| {
-				let _ = Self::_close_position(who, *position_id, None);
+		<PositionsByTrader<T>>::iter_prefix(who).for_each(|trading_pair_position_ids| {
+			trading_pair_position_ids.iter().for_each(|position_id| {
+				if let Err(err) = Self::_close_position(who, *position_id, None) {
+					debug::native::trace!(
```
Comment (shaopengw) [inline modules/margin-protocol/src/lib.rs]: Do not log here, same apply to `_liquidity_pool_liquidate`.

File: modules/margin-protocol/src/lib.rs [hunk 8]
```diff
@@ -788,6 +825,46 @@ impl<T: Trait> Module<T> {
...
+				T::LiquidityPools::get_ask_spread(pool, position.pair)
+					.ok_or(Error::<T>::NoAskSpread)?
+					.into()
+			}
+		};
+
+		// Profit from spread + penalty
+		let price_spread = fixed_128_from_fixed_u128(price)
+			.saturating_mul(spread)
+			.saturating_mul(Fixed128::from_parts(2));
```
Comment (shaopengw) [inline modules/margin-protocol/src/lib.rs]:
```suggestion
			.saturating_mul(Fixed128::from_natural(2));
```

File: modules/margin-protocol/src/lib.rs [hunk 9]
```diff
@@ -788,6 +825,46 @@ impl<T: Trait> Module<T> {
...
+
+		let spread_profit = position
+			.leveraged_held
+			.checked_mul(&price_spread)
+			.ok_or(Error::<T>::NumOutOfBound)?;
+
+		let profit_in_usd = Self::_usd_value(position.pair.base, spread_profit)?;
```
Comment (shaopengw) [inline modules/margin-protocol/src/lib.rs]: Pls refactor these lines, the naming here is confusing: your `price_spread` is not actual spread, `spread_profit` is not profit, and `profit_in_usd` is not profit in USD either.

File: modules/margin-protocol/src/lib.rs [hunk 10]
```diff
@@ -788,6 +825,46 @@ impl<T: Trait> Module<T> {
...
+	///
+	/// Return `Ok` if closure success, or `Err` if not.
+	fn _pool_liquidate(pool: LiquidityPoolId, position_id: PositionId) -> DispatchResult {
+		let position = Self::positions(position_id).ok_or(Error::<T>::PositionNotFound)?;
+
+		let price = Self::_price(position.pair.base, position.pair.quote)?;
+		let spread = {
+			if position.leverage.is_long() {
+				T::LiquidityPools::get_bid_spread(pool, position.pair)
+					.ok_or(Error::<T>::NoAskSpread)?
```
Comment (shaopengw) [inline modules/margin-protocol/src/lib.rs]:
```suggestion
					.ok_or(Error::<T>::NoBidSpread)?
```

File: modules/margin-protocol/src/tests.rs [hunk 11]
```diff
@@ -894,6 +894,46 @@ fn liquidity_pool_margin_call_and_become_safe_work() {
...
+			assert_noop!(
+				MarginProtocol::liquidity_pool_liquidate(Origin::signed(BOB), MOCK_POOL),
+				Error::<Runtime>::NotReachedRiskThreshold
+			);
+
+			// ENP 100% == 100%, unsafe
+			LiquidityPoolENPThreshold::put(risk_threshold(0, 100));
+			assert_ok!(MarginProtocol::liquidity_pool_liquidate(Origin::signed(BOB), MOCK_POOL));
+			let event = TestEvent::margin_protocol(RawEvent::LiquidityPoolLiquidated(MOCK_POOL));
+			assert!(System::events().iter().any(|record| record.event == event));
```
Comment (shaopengw) [inline modules/margin-protocol/src/tests.rs]: Need to check treasury balance.

File: modules/traits/src/lib.rs [hunk 12]
```diff
@@ -55,3 +55,7 @@ pub trait MarginProtocolLiquidityPools<AccountId>: LiquidityPools<AccountId> {
 		leveraged_amount: Balance,
 	) -> bool;
 }
+
+pub trait Treasry<AccountId> {
```
Comment (shaopengw) [inline modules/traits/src/lib.rs]:
```suggestion
pub trait Treasury<AccountId> {
```

File: modules/margin-protocol/src/lib.rs [hunk 13]
```diff
@@ -796,6 +820,51 @@ impl<T: Trait> Module<T> {
...
+		let realized = cmp::min(
+			<T::LiquidityPools as LiquidityPools<T::AccountId>>::liquidity(position.pool),
+			u128_from_fixed_128(sub_amount),
+		);
+		<T::LiquidityPools as LiquidityPools<T::AccountId>>::withdraw_liquidity(
+			&Self::account_id(),
+			position.pool,
+			realized,
+		)?;
+		<Balances<T>>::mutate(T::Treasury::account_id(), |b| *b += realized);
```
Comment (shaopengw) [inline modules/margin-protocol/src/lib.rs]: Treasury shouldn't have a balance in margin protocol, `Balances` storage is only for bookkeeping of traders.

File: modules/margin-protocol/src/lib.rs [hunk 14]
```diff
@@ -796,6 +820,51 @@ impl<T: Trait> Module<T> {
...
+		let penalty = spread_profit_in_usd;
+		let sub_amount = spread_profit_in_usd
+			.checked_add(&penalty)
+			.ok_or(Error::<T>::NumOutOfBound)?;
+
+		Self::_close_position(&position.owner, position_id, None)?;
+
+		let realized = cmp::min(
+			<T::LiquidityPools as LiquidityPools<T::AccountId>>::liquidity(position.pool),
+			u128_from_fixed_128(sub_amount),
```
Comment (shaopengw) [inline modules/margin-protocol/src/lib.rs]:
```suggestion
			u128_from_fixed_128(sub_amount.saturating_abs()),
```

sub_amount could be negative.

**Pull request merged**